test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc

```diff
@@ -0,0 +1,208 @@
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake_step_controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepControllerConfigFactory, ValidateConfigWithValidConfigReturnsOk) {
+  FakeStepControllerConfig config;
+  config.set_fixed_rps_value(1);
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake_step_controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepController, GetCurrentCommandLineOptionsReturnsRpsFromConfig) {
+  FakeStepControllerConfig config;
+  const int kExpectedValue = 5678;
+  config.set_fixed_rps_value(kExpectedValue);
+  FakeStepController step_controller(config, CommandLineOptions());
+  EXPECT_EQ(step_controller.GetCurrentCommandLineOptions().value().requests_per_second().value(),
+            kExpectedValue);
+}
+
+TEST(FakeStepController, IsConvergedInitiallyReturnsFalse) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsFalseAfterBenchmarkResultWithoutPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsTrueAfterBenchmarkResultWithPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  MetricEvaluation* evaluation = benchmark_result.mutable_metric_evaluations()->Add();
+  evaluation->set_threshold_score(1.0);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_TRUE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsDoomedReturnsFalseAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  EXPECT_FALSE(step_controller.IsDoomed(doomed_reason));
+}
+
+TEST(FakeStepController, IsDoomedDoesNotWriteDoomedReasonAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason = "untouched";
```

Comment:
I got confused by the same thing as William.  I think this would be clearer, and more accurate to the API of the call if we set this to "", and check that it's still empty after IsDoomed is called.

Reply:
One of the potential bugs this test is guarding against is IsDoomed accidentally writing "" to the variable instead of leaving it untouched, so we would want to fail if it started to do that.

What do you think of this clarified variable name and value?

Reply:
Hmm actually I see what you mean: the caller isn't supposed to have anything important in the variable ahead of time, since it might get overwritten. 

The behavior of not writing the variable when IsDoomed returns false is really not very valuable at all. Maybe the behavior when returning false should be undefined, or IsDoomed should write the variable unconditionally, writing "" if it's returning false.

Reply:
Opened #482 to rewrite IsDoomed to return absl::Status. The only reason for this weird contract is we had not brought in absl::Status at the time.

Meanwhile the test as it currently stands should explicitly test the ill-conceived spec of this function.

What do you think?

Reply:
Sounds good.  Thanks for taking a look at what could be done.  I like the Status solution.